### ISSUE TITLE
Exception in ChannelSession ReleaseSemaphore

### ISSUE THREAD

**Issue by fabricedupre** [sshnet/SSH.NET#400]
Hello,

I am using latest stable release 2016.1.0 on Windows 10 with Visual Studio 2017 within a desktop application.

ReleaseSemaphore sometimes throws a SystemNullReference exception. To circumvent that issue I had to modify the function to:

        private void ReleaseSemaphore()
        {
            if (Interlocked.CompareExchange(ref _sessionSemaphoreObtained, 0, 1) == 1)
            {
                    SemaphoreLight sem = SessionSemaphore;

                    if (sem != null)
                        sem.Release();
                    else
                        System.Diagnostics.Trace.WriteLine("*** SessionSemaphore NULL");
            }
        }

Digging a little bit, I found that in fact the issue came from Channel SessionSemaphore property. I also had to change it a little bit as below:

        protected SemaphoreLight SessionSemaphore
        {
            get
            {
                    ISession s = _session;

                    if (s != null)
                        return s.SessionSemaphore;
                
                    System.Diagnostics.Trace.WriteLine("*** session NULL");
                    return null;
            }
        }

Running my app with those changes, I sometimes get the below trace output, but no more exception:
*** session NULL
*** SessionSemaphore NULL

**Comment by marcellemke** [comment]
I occasionally had the same when I was calling the following code a lot of times (compiled SSH.Net in Debug mode as well!):
```
using (var cmd = sshclient.CreateCommand(cmdStr))
{
	var result = cmd.Execute();
}
```
As I figured out, ChannelSession.Close() has been called a second time when the exception occured.

In Session.OnChannelCloseReceived (on message handling thread), different handlers will be processed.
When exception happens, SshCommand.Channel_Closed() is being processed before Channel.OnClose(). So in my code cmd.Execute() finishes on main thread and cmd.Dispose() is being called. After Channel.Dispose() has been called in SshCommand.Dispose() (setting _session in ChannelSession to null), Channel.OnClose() is being started on the message handling thread => ReleaseSemaphore() will fail (NullReferenceException in Semaphore.get)! 

(In the screenshot I made, the main thread not only disposed the SshCommand where the exception occured, but it already created next SshCommand)

I think the fix made by fabricedupre is fine: Channel.Close()/ChannelSession.Close() are being handled at least one time (on first call) correctly.
Another solution might be to insert one of the following checks in first line of Close() methods - that won't provide full thread safety like fabricedupre' does, but those handlers in Session.OnChannelCloseReceived should all be handled sequentially on the message handling thread anyway:
```
if (!this.IsOpen) return;
if (this.IsDisposed) return;
```
![renci_except](https://user-images.githubusercontent.com/18190571/95912762-89f07680-0da3-11eb-9a47-61d8e0749917.png)
